File: Sources/ExcalidrawRender/SceneRenderer.swift
import CoreGraphics
import ExcalidrawGeometry
import ExcalidrawMath
import ExcalidrawModel
import Foundation
import FreehandKit
import RoughKit

/// Maps scene coordinates to view coordinates given the current pan/zoom.
///
/// Single source of truth for the canvas transform: render, hit-testing, and
/// snapping all derive from a `Viewport`. Mirrors the scroll/zoom model in
/// `packages/excalidraw/renderer/staticScene.ts`.
public struct Viewport: Equatable, Sendable {
    public var scrollX: Double
    public var scrollY: Double
    public var zoom: Double

    public init(scrollX: Double = 0, scrollY: Double = 0, zoom: Double = 1) {
        self.scrollX = scrollX
        self.scrollY = scrollY
        self.zoom = zoom
    }

    public func sceneToView(_ point: Point) -> Point {
        Point((point.x + scrollX) * zoom, (point.y + scrollY) * zoom)
    }

    public func viewToScene(_ point: Point) -> Point {
        Point(point.x / zoom - scrollX, point.y / zoom - scrollY)
    }

    /// Affine transform applied to the static canvas before drawing elements.
    public var affineTransform: CGAffineTransform {
        CGAffineTransform(scaleX: zoom, y: zoom)
            .translatedBy(x: scrollX, y: scrollY)
    }
}

public enum ExcalidrawRender {
    public static let zoomRange: ClosedRange<Double> = 0.1 ... 30
}

/// Renders a scene's static content (background, grid, elements) into a
/// `CGContext`. Assumes a y-down context (SwiftUI `Canvas` via `withCGContext`,
/// a UIKit view, or `UIGraphicsImageRenderer`). Ports the flow of
/// `packages/excalidraw/renderer/staticScene.ts`.
public final class SceneRenderer: SceneRendering {
    private let shapeCache: ShapeCache
    private let imageDecoder: ImageDecoder

    public init(shapeCache: ShapeCache = ShapeCache(), imageDecoder: ImageDecoder = ImageDecoder()) {
        self.shapeCache = shapeCache
        self.imageDecoder = imageDecoder
    }

    private var theme: Theme = .light
    /// Extra scene-unit margin so wide strokes near the viewport edge aren't culled.
    private let cullingMargin = 100.0

    /// Parse a colour string and apply the current theme filter.
    private func themed(_ string: String) -> CGColor {
        let base = ColorParser.cgColor(string) ?? CGColor(red: 0, green: 0, blue: 0, alpha: 1)
        return ThemeFilter.apply(base, theme: theme)
    }

    /// Render `scene`. Pass `clip` (a scene-space rectangle, e.g. a
    /// `DirtyRegion`) to repaint only that region for incremental redraw;
    /// `nil` paints the whole viewport.
    public func render(
        _ scene: Scene, in ctx: CGContext, viewport: Viewport, size: CGSize,
        theme: Theme = .light, clip: BoundingBox? = nil,
        skipping: Set<String> = [], fillBackground: Bool = true
    ) {
        self.theme = theme
        ctx.setShouldAntialias(true)
        ctx.setAllowsAntialiasing(true)
        ctx.interpolationQuality = .high
        // Background + grid belong to the static layer; the dynamic overlay pass
        // (`fillBackground: false`) composites onto an already-painted image.
        if fillBackground {
            let background: CGColor = if let userBackground = scene.appState.viewBackgroundColor {
                themed(userBackground)
            } else {
                ColorParser.cgColor(theme == .dark ? "#121212" : "#ffffff")
                    ?? CGColor(red: 1, green: 1, blue: 1, alpha: 1)
            }
            ctx.setFillColor(background)
            ctx.fill(CGRect(origin: .zero, size: size))
        }

        ctx.saveGState()
        ctx.scaleBy(x: viewport.zoom, y: viewport.zoom)
        ctx.translateBy(x: viewport.scrollX, y: viewport.scrollY)

        // Incremental redraw: restrict drawing to the dirty region.
        if let clip {
            ctx.clip(to: CGRect(x: clip.minX, y: clip.minY, width: clip.width, height: clip.height))
        }

        if fillBackground, scene.appState.gridModeEnabled == true {
            drawGrid(in: ctx, viewport: viewport, size: size)
        }

        // Frame bounds in scene coordinates, for clipping their children, plus
        // every element's bounds so bound text can be centered in its container.
        var frameBounds: [String: CGRect] = [:]
        var containerBounds: [String: CGRect] = [:]
        for element in scene.visibleElements {
            let b = element.base
            let rect = CGRect(x: b.x, y: b.y, width: b.width, height: b.height)
            containerBounds[element.id] = rect
            if isFrame(element) { frameBounds[element.id] = rect }
        }

        // Cull off-screen elements. Frame *bounds* are kept above regardless, so
        // an on-screen child still clips to a frame whose border is off-screen.
        let topLeft = viewport.viewToScene(Point(0, 0))
        let bottomRight = viewport.viewToScene(Point(size.width, size.height))
        var visibleRegion = BoundingBox(
            minX: topLeft.x, minY: topLeft.y, maxX: bottomRight.x, maxY: bottomRight.y
        )
        // When repainting a dirty region, only consider elements within it.
        if let clip {
            visibleRegion = BoundingBox(
                minX: max(visibleRegion.minX, clip.minX), minY: max(visibleRegion.minY, clip.minY),
                maxX: min(visibleRegion.maxX, clip.maxX), maxY: min(visibleRegion.maxY, clip.maxY)
            )
        }
        for element in Culling.visible(scene.visibleElements, in: visibleRegion, margin: cullingMargin)
            where !skipping.contains(element.id) {
            drawElement(
                element,
                in: ctx,
                files: scene.files,
                frameBounds: frameBounds,
                containerBounds: containerBounds
            )
        }
        ctx.restoreGState()
    }

    private func isFrame(_ element: ExcalidrawElement) -> Bool {
        switch element.kind {
        case .frame, .magicframe: true
        default: false
        }
    }

    private func drawElement(
        _ element: ExcalidrawElement, in ctx: CGContext,
        files: [String: BinaryFileData], frameBounds: [String: CGRect],
        containerBounds: [String: CGRect] = [:]
    ) {
        let base = element.base
        ctx.saveGState()
        defer { ctx.restoreGState() }

        if isFrame(element) {
            drawFrame(element, in: ctx)
            return
        }
        // Clip elements that belong to a frame to that frame's bounds.
        if let frameId = base.frameId, let rect = frameBounds[frameId] {
            ctx.clip(to: rect)
        }

        ctx.translateBy(x: base.x, y: base.y)
        // Rotate about the element's local centre.
        let cx = base.width / 2, cy = base.height / 2
        if base.angle != 0 {
            ctx.translateBy(x: cx, y: cy)
            ctx.rotate(by: CGFloat(base.angle))
            ctx.translateBy(x: -cx, y: -cy)
        }
        ctx.setAlpha(CGFloat(base.opacity / 100))

        switch element.kind {
        case let .text(text):
            let color = themed(base.strokeColor)
            // Centre text bound to a container (e.g. a sticky note) within it.
            if let containerId = text.containerId, let rect = containerBounds[containerId] {
                let size = TextLayout.measure(text)
                let originX = rect.midX - size.width / 2
                let originY = rect.midY - size.height / 2
                ctx.translateBy(x: originX - base.x, y: originY - base.y)
            }
            TextLayout.draw(text, base: base, in: ctx, color: color)
        case let .image(image):
            drawImage(image, base: base, in: ctx, files: files)
        case let .freedraw(props):
            drawFreedraw(props, base: base, in: ctx)
        case let .arrow(props):
            if let drawable = shapeCache.drawable(for: element) {
                drawDrawable(drawable, base: base, in: ctx)
            }
            drawArrowheads(props, base: base, in: ctx)
        case .embeddable, .iframe:
            drawEmbeddablePlaceholder(base: base, in: ctx)
        default:
            if let drawable = shapeCache.drawable(for: element) {
                drawDrawable(drawable, base: base, in: ctx)
            }
        }
    }

    /// Placeholder for embeddable/iframe elements (full web embedding is a UI
    /// concern): a rounded, filled rect with a diagonal cross so the element is
    /// visible and selectable. Drawn in element-local coordinates.
    private func drawEmbeddablePlaceholder(base: BaseProperties, in ctx: CGContext) {
        let rect = CGRect(x: 0, y: 0, width: base.width, height: base.height)
        let path = CGPath(roundedRect: rect, cornerWidth: 8, cornerHeight: 8, transform: nil)
        ctx.addPath(path)
        ctx.setFillColor(themed(base.backgroundColor == "transparent" ? "#f1f3f5" : base.backgroundColor))
        ctx.fillPath()
        ctx.addPath(path)
        ctx.setStrokeColor(themed(base.strokeColor))
        ctx.setLineWidth(max(base.strokeWidth, 1))
        ctx.strokePath()
        // Diagonal cross to signal embedded content.
        ctx.move(to: CGRect(x: 0, y: 0, width: base.width, height: base.height).origin)
        ctx.addLine(to: CGPoint(x: base.width, y: base.height))
        ctx.move(to: CGPoint(x: base.width, y: 0))
        ctx.addLine(to: CGPoint(x: 0, y: base.height))
        ctx.strokePath()
    }

    private func drawFrame(_ element: ExcalidrawElement, in ctx: CGContext) {
        let base = element.base
        let border = ThemeFilter.apply(CGColor(red: 0.6, green: 0.6, blue: 0.6, alpha: 1), theme: theme)
        let rect = CGRect(x: base.x, y: base.y, width: base.width, height: base.height)
        let path = CGPath(roundedRect: rect, cornerWidth: 8, cornerHeight: 8, transform: nil)
        ctx.addPath(path)
        ctx.setStrokeColor(border)
        ctx.setLineWidth(1)
        ctx.setLineDash(phase: 0, lengths: [])
        ctx.strokePath()

        // Name label just above the top-left corner.
        if case let .frame(name) = element.kind, let name, !name.isEmpty {
            ctx.saveGState()
            ctx.translateBy(x: base.x, y: base.y - 18)
            let text = TextProperties(fontSize: 12, text: name)
            TextLayout.draw(text, base: base, in: ctx, color: border)
            ctx.restoreGState()
        }
    }

    private func drawArrowheads(_ props: ArrowProperties, base: BaseProperties, in ctx: CGContext) {
        let pts = props.points
        guard pts.count >= 2 else { return }
        let stroke = themed(base.strokeColor)
        if let head = props.endArrowhead {
            drawArrowhead(
                at: pts[pts.count - 1], from: pts[pts.count - 2],
                head: head, base: base, color: stroke, in: ctx
            )
        }
        if let head = props.startArrowhead {
            drawArrowhead(at: pts[0], from: pts[1], head: head, base: base, color: stroke, in: ctx)
        }
    }

    private func drawArrowhead(
        at tip: Point, from prev: Point, head: Arrowhead, base: BaseProperties, color: CGColor, in ctx: CGContext
    ) {
        let dir = Vector(from: tip, origin: prev).normalized() // points toward the tip
        guard dir.magnitude > 0 else { return }
        let segLength = tip.distance(to: prev)
        let size = Swift.min(20, segLength * 0.5) + base.strokeWidth
        let angle = 25.0 * .pi / 180
        // Two barbs, rotating the reverse direction by ±angle.
        let back = Vector(-dir.u, -dir.v)
        func rotated(_ v: Vector, by a: Double) -> Vector {
            Vector(v.u * cos(a) - v.v * sin(a), v.u * sin(a) + v.v * cos(a))
        }
        let b1 = rotated(back, by: angle).scaled(by: size)
        let b2 = rotated(back, by: -angle).scaled(by: size)
        let p1 = Point(tip.x + b1.u, tip.y + b1.v)
        let p2 = Point(tip.x + b2.u, tip.y + b2.v)

        let filled = head == .triangle || head == .diamond
        let path = CGMutablePath()
        path.move(to: CGPoint(x: p1.x, y: p1.y))
        path.addLine(to: CGPoint(x: tip.x, y: tip.y))
        path.addLine(to: CGPoint(x: p2.x, y: p2.y))
        if filled { path.closeSubpath() }
        ctx.addPath(path)
        if filled {
            ctx.setFillColor(color)
            ctx.fillPath()
        } else {
            ctx.setStrokeColor(color)
            ctx.setLineWidth(CGFloat(base.strokeWidth))
            ctx.setLineCap(.round)
            ctx.setLineDash(phase: 0, lengths: [])
            ctx.strokePath()
        }
    }

    private func drawDrawable(_ drawable: Drawable, base: BaseProperties, in ctx: CGContext) {
        let strokeColor = themed(base.strokeColor)
        let fillColor = drawable.options.fill.map { themed($0) }

        for set in drawable.sets {
            let path = RoughPath.cgPath(from: set.ops)
            switch set.type {
            case .fillPath:
                if let fillColor {
                    ctx.addPath(path)
                    ctx.setFillColor(fillColor)
                    ctx.fillPath()
                }
            case .fillSketch:
                if let fillColor {
                    ctx.addPath(path)
                    ctx.setStrokeColor(fillColor)
                    let weight = drawable.options.fillWeight > 0 ? drawable.options.fillWeight : base.strokeWidth / 2
                    ctx.setLineWidth(CGFloat(weight))
                    ctx.setLineDash(phase: 0, lengths: [])
                    ctx.strokePath()
                }
            case .path:
                ctx.addPath(path)
                ctx.setStrokeColor(strokeColor)
                ctx.setLineWidth(CGFloat(base.strokeWidth))
                ctx.setLineCap(.round)
                ctx.setLineJoin(.round)
                applyDash(drawable.options.strokeLineDash, in: ctx)
                ctx.strokePath()
            }
        }
    }

    private func drawFreedraw(_ props: FreedrawProperties, base: BaseProperties, in ctx: CGContext) {
        guard !props.points.isEmpty else { return }
        let strokeColor = themed(base.strokeColor)
        let inputs = props.points.enumerated().map { index, p in
            FreehandPoint(x: p.x, y: p.y, pressure: index < props.pressures.count ? props.pressures[index] : 0.5)
        }
        let options = FreehandOptions(strokeWidth: base.strokeWidth, simulatePressure: props.simulatePressure)
        let outline = FreehandKit.strokeOutline(inputs, options: options)
        guard let first = outline.first, outline.count > 2 else { return }
        let path = CGMutablePath()
        path.move(to: CGPoint(x: first.x, y: first.y))
        for p in outline.dropFirst() {
            path.addLine(to: CGPoint(x: p.x, y: p.y))
        }
        path.closeSubpath()
        ctx.addPath(path)
        ctx.setFillColor(strokeColor) // perfect-freehand strokes are filled outlines
        ctx.fillPath()
    }

    private func drawImage(
        _ image: ImageProperties, base: BaseProperties, in ctx: CGContext, files: [String: BinaryFileData]
    ) {
        guard let fileId = image.fileId, let file = files[fileId],
              let fullImage = imageDecoder.image(fileId: fileId, dataURL: file.dataURL) else { return }
        // Honor a crop: draw only the cropped sub-region of the source, scaled
        // to fill the element rect. Crop is in natural-pixel, top-left coords.
        let cgImage: CGImage
        if let crop = image.crop {
            let cropRect = CGRect(x: crop.x, y: crop.y, width: crop.width, height: crop.height)
            cgImage = fullImage.cropping(to: cropRect) ?? fullImage
        } else {
            cgImage = fullImage
        }
        // CGContext.draw flips images in a y-down context; flip back so it is upright.
        let rect = CGRect(x: 0, y: 0, width: base.width, height: base.height)
        ctx.saveGState()
        ctx.translateBy(x: 0, y: base.height)
        ctx.scaleBy(x: 1, y: -1)
        ctx.draw(cgImage, in: rect)
        ctx.restoreGState()
    }

    private func drawGrid(in ctx: CGContext, viewport: Viewport, size: CGSize) {
        let gridSize = 20.0
        ctx.setStrokeColor(CGColor(red: 0, green: 0, blue: 0, alpha: 0.1))
        ctx.setLineWidth(1 / viewport.zoom)
        let path = CGMutablePath()
        // Visible scene bounds.
        let topLeft = viewport.viewToScene(Point(0, 0))
        let bottomRight = viewport.viewToScene(Point(size.width, size.height))
        var x = (topLeft.x / gridSize).rounded(.down) * gridSize
        while x <= bottomRight.x {
            path.move(to: CGPoint(x: x, y: topLeft.y))
            path.addLine(to: CGPoint(x: x, y: bottomRight.y))
            x += gridSize
        }
        var y = (topLeft.y / gridSize).rounded(.down) * gridSize
        while y <= bottomRight.y {
            path.move(to: CGPoint(x: topLeft.x, y: y))
            path.addLine(to: CGPoint(x: bottomRight.x, y: y))
            y += gridSize
        }
        ctx.addPath(path)
        ctx.strokePath()
    }

    private func applyDash(_ dash: [Double]?, in ctx: CGContext) {
        if let dash, !dash.isEmpty {
            ctx.setLineDash(phase: 0, lengths: dash.map { CGFloat($0) })
        } else {
            ctx.setLineDash(phase: 0, lengths: [])
        }
    }
}

File: web/packages/render/src/scene-renderer.ts
import { BoundingBox, cullVisible } from "@xs/geometry";
import type { ExcalidrawElement, Scene, TextElement } from "@xs/model";
import { viewBackgroundColor } from "@xs/model";
import { getStroke } from "perfect-freehand";
import { type PathSink, opsToPath } from "./drawable-path.js";
import { elementDrawable } from "./element-drawable.js";
import { buildRoughOptions } from "./rough-options.js";
import type { Viewport } from "./viewport.js";

/** The subset of `CanvasRenderingContext2D` the renderer uses. */
export interface RenderContext extends PathSink {
  save(): void;
  restore(): void;
  translate(x: number, y: number): void;
  scale(x: number, y: number): void;
  rotate(angle: number): void;
  closePath(): void;
  fill(): void;
  stroke(): void;
  fillRect(x: number, y: number, w: number, h: number): void;
  strokeRect(x: number, y: number, w: number, h: number): void;
  arc(x: number, y: number, radius: number, startAngle: number, endAngle: number): void;
  setLineDash(segments: number[]): void;
  fillText(text: string, x: number, y: number): void;
  fillStyle: string;
  strokeStyle: string;
  lineWidth: number;
  lineCap: CanvasLineCap;
  lineJoin: CanvasLineJoin;
  globalAlpha: number;
  font: string;
}

export type Theme = "light" | "dark";

export interface RenderOptions {
  viewport: Viewport;
  width: number;
  height: number;
  theme?: Theme;
  gridSize?: number;
}

function backgroundColor(scene: Scene, theme: Theme): string {
  return viewBackgroundColor(scene.appState) ?? (theme === "dark" ? "#121212" : "#ffffff");
}

function drawDrawable(ctx: RenderContext, el: ExcalidrawElement): void {
  const options = buildRoughOptions(el);
  const drawable = elementDrawable(el, options);
  if (drawable === null) return;
  const fill = drawable.options.fill;
  for (const set of drawable.sets) {
    opsToPath(set.ops, ctx);
    if (set.type === "fillPath") {
      if (fill) {
        ctx.fillStyle = fill;
        ctx.fill();
      }
    } else if (set.type === "fillSketch") {
      if (fill) {
        ctx.strokeStyle = fill;
        ctx.lineWidth =
          drawable.options.fillWeight && drawable.options.fillWeight > 0
            ? drawable.options.fillWeight
            : el.strokeWidth / 2;
        ctx.setLineDash([]);
        ctx.stroke();
      }
    } else {
      ctx.strokeStyle = el.strokeColor;
      ctx.lineWidth = el.strokeWidth;
      ctx.lineCap = "round";
      ctx.lineJoin = "round";
      ctx.setLineDash(drawable.options.strokeLineDash ?? []);
      ctx.stroke();
    }
  }
}

function drawFreedraw(
  ctx: RenderContext,
  el: Extract<ExcalidrawElement, { type: "freedraw" }>,
): void {
  const inputs = el.points.map(
    (p, i) => [p[0], p[1], el.pressures[i] ?? 0.5] as [number, number, number],
  );
  const outline = getStroke(inputs, {
    size: Math.max(el.strokeWidth, 1) * 4.25,
    simulatePressure: el.simulatePressure,
  });
  if (outline.length < 3) return;
  ctx.beginPath();
  ctx.moveTo(outline[0]![0]!, outline[0]![1]!);
  for (let i = 1; i < outline.length; i++) ctx.lineTo(outline[i]![0]!, outline[i]![1]!);
  ctx.closePath();
  ctx.fillStyle = el.strokeColor;
  ctx.fill();
}

function drawText(ctx: RenderContext, el: TextElement): void {
  ctx.fillStyle = el.strokeColor;
  ctx.font = `${el.fontSize}px sans-serif`;
  const lineHeight = el.fontSize * el.lineHeight;
  const lines = el.text.split("\n");
  for (let i = 0; i < lines.length; i++) {
    if (lines[i]!.length === 0) continue;
    ctx.fillText(lines[i]!, 0, lineHeight * i + el.fontSize * 0.8);
  }
}

function drawFrame(ctx: RenderContext, el: ExcalidrawElement): void {
  ctx.strokeStyle = "#999999";
  ctx.lineWidth = 1;
  ctx.setLineDash([]);
  ctx.beginPath();
  ctx.moveTo(0, 0);
  ctx.lineTo(el.width, 0);
  ctx.lineTo(el.width, el.height);
  ctx.lineTo(0, el.height);
  ctx.closePath();
  ctx.stroke();
}

function visibleRegion(opts: RenderOptions): BoundingBox {
  const { viewport: v, width, height } = opts;
  return new BoundingBox(
    -v.scrollX,
    -v.scrollY,
    width / v.zoom - v.scrollX,
    height / v.zoom - v.scrollY,
  );
}

/**
 * Render a scene's static content (background, grid, elements) into a 2D
 * context. Assumes a y-down context. (parity: SceneRenderer.swift)
 */
export function renderScene(ctx: RenderContext, scene: Scene, opts: RenderOptions): void {
  const theme = opts.theme ?? "light";
  ctx.fillStyle = backgroundColor(scene, theme);
  ctx.fillRect(0, 0, opts.width, opts.height);

  ctx.save();
  ctx.scale(opts.viewport.zoom, opts.viewport.zoom);
  ctx.translate(opts.viewport.scrollX, opts.viewport.scrollY);

  const region = visibleRegion(opts);
  const elements = cullVisible(scene.visibleElements, region, 100);

  for (const el of elements) {
    ctx.save();
    ctx.globalAlpha = el.opacity / 100;
    ctx.translate(el.x, el.y);
    if (el.angle !== 0) {
      ctx.translate(el.width / 2, el.height / 2);
      ctx.rotate(el.angle);
      ctx.translate(-el.width / 2, -el.height / 2);
    }
    switch (el.type) {
      case "text":
        drawText(ctx, el);
        break;
      case "freedraw":
        drawFreedraw(ctx, el);
        break;
      case "frame":
      case "magicframe":
        drawFrame(ctx, el);
        break;
      case "image":
        break; // image bitmaps are drawn by the host
      default:
        drawDrawable(ctx, el);
        break;
    }
    ctx.restore();
  }
  ctx.restore();
}
